File: Contest/Biweekly Contest 84/Minimum Replacements to Sort the Array/MinReplacements.kt
class Solution {
    fun minimumReplacement(nums: IntArray): Long {
        val n = nums.size
        if (n == 1) return 0
        
        var ans: Long = 0
        var i = n - 2
        var prev = nums[n-1]
        while (i >= 0) {
            val curr = nums[i]
            
            if (curr > prev) {
                if (curr % prev == 0) {
                    ans += curr / prev - 1
                } else {
                    ans += curr / prev
                    prev = (curr - (curr / prev - 1) * prev) / 2
                }
            } else {
                prev = curr
            }
            
            i--
        }
        
        return ans
    }
}

File: Problems/Algorithms/2366. Minimum Replacements to Sort the Array/min-replacements.cpp
#include <vector>
#include <cmath>

using namespace std;

class Solution {
public:
    long long minimumReplacement(vector<int>& nums) {
        long long ans = 0;
        int n = nums.size();

        for (int i = n-2; i >= 0; i--) {
            if (nums[i] <= nums[i+1]) {
                continue;
            }

            long long count = ceil((nums[i] + nums[i+1] - 1LL) / nums[i+1]);
            ans += count - 1;

            nums[i] = nums[i] / count;
        }

        return ans;
    }
};
